File: client-java/instrumentation/src/main/java/org/evomaster/clientJava/instrumentation/TargetInfo.java
package org.evomaster.clientJava.instrumentation;

import java.io.Serializable;

/**
 * This represents the same data as in TargetInfoDto.
 * Here is replicated to have a clear distinction on how
 * such data is used
 */
public class TargetInfo implements Serializable{

    public final Integer mappedId;

    public final String descriptiveId;

    /**
     * heuristic [0,1], where 1 means covered
     */
    public final Double value;

    /**
     * Can be negative if target was never reached.
     * But this means that {@code value} must be 0
     */
    public final Integer actionIndex;

    public TargetInfo(Integer mappedId, String descriptiveId, Double value, Integer actionIndex) {
        this.mappedId = mappedId;
        this.descriptiveId = descriptiveId;
        this.value = value;
        this.actionIndex = actionIndex;
    }

    public static TargetInfo notReached(int theID){
        return new TargetInfo(theID, null, 0d, -1);
    }

    public TargetInfo withMappedId(int theID){
        if(mappedId != null){
            throw new IllegalArgumentException("Id already existing");
        }
        return new TargetInfo(theID, descriptiveId, value, actionIndex);
    }

    public TargetInfo withNoDescriptiveId(){
        return new TargetInfo(mappedId, null, value, actionIndex);
    }
}

File: client-js/evomaster-client-js/src/instrumentation/TargetInfo.ts
/**
 * This represents the same data as in TargetInfoDto.
 * Here is replicated to have a clear distinction on how
 * such data is used
 */
export default class TargetInfo {

    public static notReached(theID: number): TargetInfo {
        return new TargetInfo(theID, null, 0.0, -1);
    }

    public readonly mappedId: number;

    public readonly descriptiveId: string;

    /**
     * heuristic [0,1], where 1 means covered
     */
    public readonly value: number;

    /**
     * Can be negative if target was never reached.
     * But this means that {@code value} must be 0
     */
    public readonly actionIndex: number;

    constructor(mappedId: number, descriptiveId: string, value: number, actionIndex: number) {
        this.mappedId = mappedId;
        this.descriptiveId = descriptiveId;
        this.value = value;
        this.actionIndex = actionIndex;
    }

    public withMappedId(theID: number): TargetInfo {

        if (this.mappedId) {
            throw new Error("Id already existing");
        }
        return new TargetInfo(theID, this.descriptiveId, this.value, this.actionIndex);
    }

    public withNoDescriptiveId(): TargetInfo {
        return new TargetInfo(this.mappedId, null, this.value, this.actionIndex);
    }
}
